README.md
# Example

## AES

The example of AES-256-CBC, others are similar.

```go
package main

import (
	"bytes"
	"fmt"

	"github.com/garvenc/go-crypt" // Package name is crypt.
)

func main() {
	input := "123456"
	key := bytes.Repeat([]byte{'a'}, crypt.Aes256KeySize)
	iv := bytes.Repeat([]byte{'b'}, crypt.AesIvSize)

	encrypter := crypt.NewAesCbcEncrypter(key, iv, crypt.NewPkcs7Padding(crypt.AesBlockSize))
	decrypter := crypt.NewAesCbcDecrypter(key, iv, crypt.NewPkcs7Padding(crypt.AesBlockSize))
	enc, err := encrypter.Encrypt([]byte(input))
	if err != nil {
		fmt.Println("encrypt:", err)
		return
	}
	fmt.Println("encrypt result:", enc)
	dec, err := decrypter.Decrypt(enc)
	if err != nil {
		fmt.Println("decrypt:", err)
		return
	}
	fmt.Println("decrypt result:", dec)
	fmt.Println("decrypt result string:", string(dec))
}
```

Output:

```
encrypt result: [91 14 236 68 51 250 203 219 122 12 215 110 44 233 203 201]
decrypt result: [49 50 51 52 53 54]
decrypt result string: 123456
```

## RSA

```go
package main

import (
	"fmt"

	"github.com/garvenc/go-crypt" // Package name is crypt.
)

func main() {
	input := "123456"
	bits := 1024
	e := 65537

	pri := crypt.NewRsaPrivate(bits, e)
	n := pri.GetNBytes()
	fmt.Println("n:", n) // Each execution result is different.
	pub := crypt.NewRsaPublic(n, e)
	enc, err := pub.PublicEncryptPkcs1v15([]byte(input))
	if err != nil {
		fmt.Println("public encrypt:", err)
		return
	}
	fmt.Println("public encrypt result:", enc) // Each execution result is different.
	dec, err := pri.PrivateDecryptPkcs1v15(enc)
	if err != nil {
		fmt.Println("privage decrypt:", err)
		return
	}
	fmt.Println("private decrypt result:", dec)
	fmt.Println("private decrypt result string:", string(dec))
}
```

Output:

```
n: [231 233 187 169 26 184 91 211 208 182 160 66 238 180 224 9 15 99 152 165 75 175 226 248 192 140 196 197 75 239 37 136 125 1 175 49 191 180 166 37 152 3 29 245 203 235 38 130 89 155 145 22 75 94 159 180 174 190 160 196 32 29 199 101 79 167 191 75 252 222 141 187 1 158 41 91 208 94 249 131 61 87 241 35 224 199 251 102 51 248 245 55 222 243 202 251 2 235 55 75 98 181 229 22 98 26 20 14 182 44 158 36 142 66 208 214 177 43 15 105 97 103 253 20 140 76 209 91]
public encrypt result: [96 84 95 118 182 215 53 138 124 22 113 38 222 128 203 59 21 199 127 91 211 250 216 153 146 248 82 87 15 36 73 144 157 228 63 143 182 141 132 148 106 97 18 160 42 34 185 4 243 247 234 237 135 73 155 228 188 143 83 195 177 2 24 141 158 246 115 30 150 76 65 151 5 164 158 64 89 12 242 43 155 65 26 192 86 26 147 85 16 255 1 50 53 189 38 120 8 182 26 50 35 236 205 56 188 208 68 33 177 106 110 23 114 86 151 223 200 108 40 58 2 142 85 61 246 169 78 37]
private decrypt result: [49 50 51 52 53 54]
private decrypt result string: 123456
```


HISTORY.md
# v1.0.0

* Support AES-CBC, AES-CFB, AES-CTR, AES-OFB.
* Support RSA PKCS#1 v1.5.
* Support PKCS#5 and PKCS#7 padding.

aes_test.go
package crypt

import (
	"bytes"
	"encoding/base64"
	"testing"
)

func TestAesCbc(t *testing.T) {
	key := []byte("11112222333344445555666677778888")
	iv := []byte("1234567812345678")
	data := "I love this girl! Does she?"
	result := "DC9HZuq4EOq7fO+vP2Qs2Oh9zfaA8TI/u6tHN38yvcM="

	t.Run("encrypt", func(t *testing.T) {
		encrypter := NewAesCbcEncrypter(key, iv, NewPkcs7Padding(AesBlockSize))
		enc, err := encrypter.Encrypt([]byte(data))
		if err != nil {
			t.Error("encrypt:", err)
		}
		str := base64.StdEncoding.EncodeToString(enc)
		if str != result {
			t.Error("encrypt result is wrong:", str)
		}
	})
	t.Run("decrypt", func(t *testing.T) {
		buf, _ := base64.StdEncoding.DecodeString(result)
		decrypter := NewAesCbcDecrypter(key, iv, NewPkcs7Padding(AesBlockSize))
		dec, err := decrypter.Decrypt(buf)
		if err != nil {
			t.Error("decrypt:", err)
		}
		if string(dec) != data {
			t.Error("decrypt result is wrong:", dec)
		}
	})
}

func TestAesCfb(t *testing.T) {
	key := []byte("11112222333344445555666677778888")
	iv := []byte("1234567812345678")
	data := "I love this girl! Does she?"
	result := "+HVXA7n2iUln6vXL2buTcUuN844bbH5c1XEz"

	t.Run("encrypt", func(t *testing.T) {
		encrypter := NewAesCfbEncrypter(key, iv)
		enc, err := encrypter.Crypt([]byte(data))
		if err != nil {
			t.Error("encrypt:", err)
		}
		str := base64.StdEncoding.EncodeToString(enc)
		if str != result {
			t.Error("encrypt result is wrong:", str)
		}
	})
	t.Run("decrypt", func(t *testing.T) {
		buf, _ := base64.StdEncoding.DecodeString(result)
		decrypter := NewAesCfbDecrypter(key, iv)
		dec, err := decrypter.Crypt(buf)
		if err != nil {
			t.Error("decrypt:", err)
		}
		if string(dec) != data {
			t.Error("decrypt result is wrong:", dec)
		}
	})
}

func TestAesOfb(t *testing.T) {
	key := []byte("11112222333344445555666677778888")
	iv := []byte("1234567812345678")
	data := "I love this girl! Does she?"
	result := "+HVXA7n2iUln6vXL2buTcQ/iMrzkSWcxWYEn"

	t.Run("encrypt", func(t *testing.T) {
		encrypter := NewAesOfb(key, iv)
		enc, err := encrypter.Crypt([]byte(data))
		if err != nil {
			t.Error("encrypt:", err)
		}
		str := base64.StdEncoding.EncodeToString(enc)
		if str != result {
			t.Error("encrypt result is wrong:", str)
		}
	})
	t.Run("decrypt", func(t *testing.T) {
		buf, _ := base64.StdEncoding.DecodeString(result)
		decrypter := NewAesOfb(key, iv)
		dec, err := decrypter.Crypt(buf)
		if err != nil {
			t.Error("decrypt:", err)
		}
		if string(dec) != data {
			t.Error("decrypt result is wrong:", dec)
		}
	})
}

func TestAesCtr(t *testing.T) {
	key := []byte("11112222333344445555666677778888")
	iv := []byte("1234567812345678")
	data := "I love this girl! Does she?"
	result := "+HVXA7n2iUln6vXL2buTcfv28+am206YJuzC"

	t.Run("encrypt", func(t *testing.T) {
		encrypter := NewAesCtr(key, iv)
		enc, err := encrypter.Crypt([]byte(data))
		if err != nil {
			t.Error("encrypt:", err)
		}
		str := base64.StdEncoding.EncodeToString(enc)
		if str != result {
			t.Error("encrypt result is wrong:", str)
		}
	})
	t.Run("decrypt", func(t *testing.T) {
		buf, _ := base64.StdEncoding.DecodeString(result)
		decrypter := NewAesCtr(key, iv)
		dec, err := decrypter.Crypt(buf)
		if err != nil {
			t.Error("decrypt:", err)
		}
		if string(dec) != data {
			t.Error("decrypt result is wrong:", dec)
		}
	})
}

func BenchmarkAes256CbcEncrypt1000Bytes(b *testing.B) {
	b.StopTimer()
	key := bytes.Repeat([]byte("a"), Aes256KeySize)
	iv := bytes.Repeat([]byte("b"), AesIvSize)
	data := bytes.Repeat([]byte("s"), 1000)
	encrypter := NewAesCbcEncrypter(key, iv, NewPkcs7Padding(8))
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		encrypter.Encrypt(data)
	}
}

func BenchmarkAes256CbcDecrypt1000Bytes(b *testing.B) {
	b.StopTimer()
	key := bytes.Repeat([]byte("a"), Aes256KeySize)
	iv := bytes.Repeat([]byte("b"), AesIvSize)
	data := bytes.Repeat([]byte("s"), 1000)
	encrypter := NewAesCbcEncrypter(key, iv, NewPkcs7Padding(AesBlockSize))
	decrypter := NewAesCbcDecrypter(key, iv, NewPkcs7Padding(AesBlockSize))
	enc, err := encrypter.Encrypt(data)
	if err != nil {
		b.Error("prepare enc data fail:", err)
		return
	}
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		decrypter.Decrypt(enc)
	}
}

func BenchmarkAes256CfbEncrypt1000Bytes(b *testing.B) {
	b.StopTimer()
	key := bytes.Repeat([]byte("a"), Aes256KeySize)
	iv := bytes.Repeat([]byte("b"), AesIvSize)
	data := bytes.Repeat([]byte("s"), 1000)
	encrypter := NewAesCfbEncrypter(key, iv)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		encrypter.Crypt(data)
	}
}

func BenchmarkAes256CfbDecrypt1000Bytes(b *testing.B) {
	b.StopTimer()
	key := bytes.Repeat([]byte("a"), Aes256KeySize)
	iv := bytes.Repeat([]byte("b"), AesIvSize)
	data := bytes.Repeat([]byte("s"), 1000)
	encrypter := NewAesCfbEncrypter(key, iv)
	decrypter := NewAesCfbDecrypter(key, iv)
	enc, err := encrypter.Crypt(data)
	if err != nil {
		b.Error("prepare enc data fail:", err)
		return
	}
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		decrypter.Crypt(enc)
	}
}

func BenchmarkAes256OfbCrypt1000Bytes(b *testing.B) {
	b.StopTimer()
	key := bytes.Repeat([]byte("a"), Aes256KeySize)
	iv := bytes.Repeat([]byte("b"), AesIvSize)
	data := bytes.Repeat([]byte("s"), 1000)
	crypter := NewAesOfb(key, iv)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		crypter.Crypt(data)
	}
}

func BenchmarkAes256CtrCrypt1000Bytes(b *testing.B) {
	b.StopTimer()
	key := bytes.Repeat([]byte("a"), Aes256KeySize)
	iv := bytes.Repeat([]byte("b"), AesIvSize)
	data := bytes.Repeat([]byte("s"), 1000)
	crypter := NewAesCtr(key, iv)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		crypter.Crypt(data)
	}
}


rsa_test.go
package crypt

import (
	"bytes"
	"testing"
)

func TestRsaPkcs1v15PublicEncryptAndPrivateDecrypt(t *testing.T) {
	bits, e := 1024, 65537
	pri := NewRsaPrivate(bits, e)
	pub := NewRsaPublic(pri.GetNBytes(), e)
	input := bytes.Repeat([]byte{0}, 1000)

	enc, err := pub.PublicEncryptPkcs1v15(input)
	if err != nil {
		t.Error("public encrypt err:", err)
	}
	dec, err := pri.PrivateDecryptPkcs1v15(enc)
	if err != nil {
		t.Error("private decrypt err:", err)
	}
	if !bytes.Equal(input, dec) {
		t.Error("the decrypted is not equal to the input")
	}
}

func BenchmarkNewRsaPrivate(b *testing.B) {
	for i := 0; i < b.N; i++ {
		NewRsaPrivate(1024, 65537)
	}
}

func BenchmarkNewRsaPublic(b *testing.B) {
	b.StopTimer()
	pri := NewRsaPrivate(1024, 65537)
	n := pri.GetNBytes()
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		NewRsaPublic(n, 65537)
	}
}

func BenchmarkRsa1024PublicEncryptPkcs1v15With1000Bytes(b *testing.B) {
	b.StopTimer()
	bits, e := 1024, 65537
	pri := NewRsaPrivate(bits, e)
	pub := NewRsaPublic(pri.GetNBytes(), e)
	input := bytes.Repeat([]byte{0}, 1000)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		pub.PublicEncryptPkcs1v15(input)
	}
}

func BenchmarkRsa1024PrivateDecryptPkcs1v15With1000Bytes(b *testing.B) {
	b.StopTimer()
	bits, e := 1024, 65537
	pri := NewRsaPrivate(bits, e)
	pub := NewRsaPublic(pri.GetNBytes(), e)
	input := bytes.Repeat([]byte{0}, 1000)
	enc, err := pub.PublicEncryptPkcs1v15(input)
	if err != nil {
		b.Error("public encrypt err:", err)
	}
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		pri.PrivateDecryptPkcs1v15(enc)
	}
}

func BenchmarkRsa1024GetNBytes(b *testing.B) {
	b.StopTimer()
	bits, e := 1024, 65537
	pri := NewRsaPrivate(bits, e)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		pri.GetNBytes()
	}
}


padding_test.go
package crypt

import (
	"bytes"
	"testing"
)

func TestPkcs7Padding(t *testing.T) {
	buf := []byte{1, 2, 3, 0, 0, 0}
	bufOrigin := []byte{1, 2, 3, 0, 0, 0}
	bufUnpad := bufOrigin[:3]
	bufPad := []byte{1, 2, 3, 2, 2}

	padding := NewPkcs7Padding(5)
	t.Run("Pad", func(t *testing.T) {
		result := padding.Pad(bufUnpad)
		if !bytes.Equal(result, bufPad) {
			t.Error("Pad wrong result:", result)
		}
		if !bytes.Equal(bufOrigin, buf) {
			t.Error("the origin buf has been modified unexpectedly")
		}
	})
	t.Run("Unpad", func(t *testing.T) {
		result, err := padding.Unpad(bufPad)
		if err != nil {
			t.Error("Unpad:", err)
		}
		if !bytes.Equal(result, bufUnpad) {
			t.Error("Unpad wrong result:", result)
		}
	})
}

func BenchmarkPkcs7PaddingPad(b *testing.B) {
	b.StopTimer()
	blockSize := 16 // aes block size
	padding := NewPkcs7Padding(blockSize)
	data := bytes.Repeat([]byte{0}, blockSize*10+1)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		padding.Pad(data)
	}
}

func BenchmarkPkcs7PaddingUnpad(b *testing.B) {
	b.StopTimer()
	blockSize := 16 // aes block size
	padding := NewPkcs7Padding(blockSize)
	data := bytes.Repeat([]byte{0}, blockSize*10+1)
	bufPad := padding.Pad(data)
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		padding.Unpad(bufPad)
	}
}


rsa.go
package crypt

import (
	"bytes"
	"crypto/rand"
	"crypto/rsa"
	"errors"
	"io"
	"math/big"
)

var errRsaPublicKeySize = errors.New("rsa public key size illegal")

// This is copy from crypto/rsa.
var bigOne = big.NewInt(1)

// It may has an error, call HasError to see it.
type RsaPrivate struct {
	priKey *rsa.PrivateKey
	err    error
}

// bits:
// The bit size of rsa key. It must be multiple of 8.
// The byte size of N equals to bits/8. 1024 usually, 2048 in some important cases.
//
// e:
// The public exponent E value. Usually use 65537.
//
// Can call HasError to see if it has an error.
func NewRsaPrivate(bits, e int) RsaPrivate {
	pri := RsaPrivate{}
	if bits%8 != 0 {
		pri.err = errRsaPublicKeySize
		return pri
	}
	key, err := generateMultiPrimeKey(rand.Reader, 2, bits, e) // Call it like what in crypto/rsa.GenerateKey.
	if err != nil {
		pri.err = err
		return pri
	}
	pri.priKey = key
	pri.err = nil
	return pri
}

func (p RsaPrivate) HasError() (error, bool) {
	return p.err, p.err != nil
}

// Public encrypt of PKCS#1 v1.5.
// Each bits/8-11 or less bytes will be encrypted to bits/8 bytes.
// If len(data) == 0, will return an empty buf too.
//
// The result will not share the array of data.
func (p RsaPrivate) PublicEncryptPkcs1v15(data []byte) ([]byte, error) {
	if p.err != nil {
		return nil, p.err
	}
	return publicEncryptPkcs1v15(&p.priKey.PublicKey, data)
}

// Private Decrypt of PKCS#1 v1.5.
// Each bits/8 bytes will be decrypted to bits/8-11 or less bytes.
// If len(data) == 0, will return an empty buf too.
//
// The result will not share the array of data.
func (p RsaPrivate) PrivateDecryptPkcs1v15(data []byte) ([]byte, error) {
	if p.err != nil {
		return nil, p.err
	}
	return privateDecryptPkcs1v15(p.priKey, data)
}

// If p has error, return nil.
func (p RsaPrivate) GetNBytes() []byte {
	if p.err != nil {
		return nil
	}
	return p.priKey.PublicKey.N.Bytes()
}

type RsaPublic struct {
	pubKey *rsa.PublicKey
}

// n:
// The bytes of N. The size of N is the length of the public modulus.
//
// e:
// The public exponent E value. Usually use 65537.
func NewRsaPublic(n []byte, e int) RsaPublic {
	return RsaPublic{
		pubKey: &rsa.PublicKey{
			N: new(big.Int).SetBytes(n),
			E: e,
		},
	}
}

// Public encrypt of PKCS#1 v1.5.
// Each bits/8-11 or less bytes will be encrypted to bits/8 bytes.
// If len(data) == 0, will return an empty buf too.
//
// The result will not share the array of data.
func (pub *RsaPublic) PublicEncryptPkcs1v15(data []byte) ([]byte, error) {
	return publicEncryptPkcs1v15(pub.pubKey, data)
}

func publicEncryptPkcs1v15(pub *rsa.PublicKey, data []byte) ([]byte, error) {
	// Each input must be not longer than the length of bytes of the public modulus minus 11 bytes.
	// The output is always equal to the length of bytes of the public modulus.
	eachSize := pub.N.BitLen()/8 - 11
	if eachSize <= 0 {
		return nil, errRsaPublicKeySize
	}
	var outputs [][]byte
	for encSize, leftSize, thisSize := 0, len(data), 0; leftSize > 0; {
		if leftSize > eachSize {
			thisSize = eachSize
		} else {
			thisSize = leftSize
		}
		out, err := rsa.EncryptPKCS1v15(rand.Reader, pub, data[encSize:(encSize+thisSize)])
		if err != nil {
			return nil, err
		}
		outputs = append(outputs, out)
		encSize += thisSize
		leftSize -= thisSize
	}
	return bytes.Join(outputs, nil), nil
}

func privateDecryptPkcs1v15(pri *rsa.PrivateKey, data []byte) ([]byte, error) {
	// See publicEncryptPkcs1v15. The logical procedure is reversed.
	eachSize := pri.PublicKey.N.BitLen() / 8
	if eachSize <= 0 {
		return nil, errRsaPublicKeySize
	}
	var outputs [][]byte
	for encSize, leftSize, thisSize := 0, len(data), 0; leftSize > 0; {
		if leftSize > eachSize {
			thisSize = eachSize
		} else {
			thisSize = leftSize
		}
		out, err := rsa.DecryptPKCS1v15(rand.Reader, pri, data[encSize:(encSize+thisSize)])
		if err != nil {
			return nil, err
		}
		outputs = append(outputs, out)
		encSize += thisSize
		leftSize -= thisSize
	}
	return bytes.Join(outputs, nil), nil
}

// This is copy from crypto/rsa.GenerateMultiPrimeKey, except pass an additional e param and can set priv.E as it.
// In the rsa.GenerateMultiPrimeKey, priv.E is always 65537.
func generateMultiPrimeKey(random io.Reader, nprimes int, bits int, e int) (priv *rsa.PrivateKey, err error) {
	priv = new(rsa.PrivateKey)
	priv.E = e // In crypto/rsa.GenerateMultiPrimeKey, it is always 65537.

	if nprimes < 2 {
		return nil, errors.New("crypto/rsa: GenerateMultiPrimeKey: nprimes must be >= 2")
	}

	primes := make([]*big.Int, nprimes)

NextSetOfPrimes:
	for {
		todo := bits
		// crypto/rand should set the top two bits in each prime.
		// Thus each prime has the form
		//   p_i = 2^bitlen(p_i) × 0.11... (in base 2).
		// And the product is:
		//   P = 2^todo × α
		// where α is the product of nprimes numbers of the form 0.11...
		//
		// If α < 1/2 (which can happen for nprimes > 2), we need to
		// shift todo to compensate for lost bits: the mean value of 0.11...
		// is 7/8, so todo + shift - nprimes * log2(7/8) ~= bits - 1/2
		// will give good results.
		if nprimes >= 7 {
			todo += (nprimes - 2) / 5
		}
		for i := 0; i < nprimes; i++ {
			primes[i], err = rand.Prime(random, todo/(nprimes-i))
			if err != nil {
				return nil, err
			}
			todo -= primes[i].BitLen()
		}

		// Make sure that primes is pairwise unequal.
		for i, prime := range primes {
			for j := 0; j < i; j++ {
				if prime.Cmp(primes[j]) == 0 {
					continue NextSetOfPrimes
				}
			}
		}

		n := new(big.Int).Set(bigOne)
		totient := new(big.Int).Set(bigOne)
		pminus1 := new(big.Int)
		for _, prime := range primes {
			n.Mul(n, prime)
			pminus1.Sub(prime, bigOne)
			totient.Mul(totient, pminus1)
		}
		if n.BitLen() != bits {
			// This should never happen for nprimes == 2 because
			// crypto/rand should set the top two bits in each prime.
			// For nprimes > 2 we hope it does not happen often.
			continue NextSetOfPrimes
		}

		g := new(big.Int)
		priv.D = new(big.Int)
		y := new(big.Int)
		e := big.NewInt(int64(priv.E))
		g.GCD(priv.D, y, e, totient)

		if g.Cmp(bigOne) == 0 {
			if priv.D.Sign() < 0 {
				priv.D.Add(priv.D, totient)
			}
			priv.Primes = primes
			priv.N = n

			break
		}
	}

	priv.Precompute()
	return
}


padding.go
package crypt

import (
	"errors"
)

var (
	errPaddingIsWrong = errors.New("padding is wrong")
)

type Padding interface {
	BlockSize() int
	// Must ensure the result do not use the array out of boundary of the param slice.
	// This usually means it should return a slice with a new array.
	Pad([]byte) []byte
	// The result may be part of the param slice.
	Unpad([]byte) ([]byte, error)
}

// PKCS#7 padding: See 10.3 of http://tools.ietf.org/html/rfc2315
type Pkcs7Padding struct {
	blockSize int
}

// blockSize: It is the size of bytes in a block.
func NewPkcs7Padding(blockSize int) Pkcs7Padding {
	return Pkcs7Padding{
		blockSize: blockSize,
	}
}

func (p Pkcs7Padding) BlockSize() int {
	return p.blockSize
}

func (p Pkcs7Padding) Pad(buf []byte) []byte {
	paddingSize := p.blockSize - len(buf)%p.blockSize
	result := make([]byte, len(buf)+paddingSize)
	copy(result, buf)
	for i := 0; i < paddingSize; i++ {
		result[len(buf)+i] = byte(paddingSize)
	}
	return result
}

func (p Pkcs7Padding) Unpad(buf []byte) ([]byte, error) {
	paddingSize, err := p.paddingSizeOfUnpad(buf)
	if err != nil {
		return nil, err
	}
	return buf[:len(buf)-paddingSize], nil
}

func (p Pkcs7Padding) paddingSizeOfUnpad(buf []byte) (int, error) {
	length := len(buf)
	if length <= 0 {
		return 0, errPaddingIsWrong
	}
	value := buf[length-1]
	paddingSize := int(value)
	if paddingSize > length {
		return 0, errPaddingIsWrong
	}
	for i := 2; i <= paddingSize; i++ { // The src[length-1] is already value.
		if buf[length-i] != value {
			return 0, errPaddingIsWrong
		}
	}
	return paddingSize, nil
}

// PKCS#5 padding: See 6.1.1 of http://tools.ietf.org/html/rfc2898
type Pkcs5Padding struct {
	pkcs7 Pkcs7Padding
}

func NewPkcs5Padding() Pkcs5Padding {
	return Pkcs5Padding{
		pkcs7: NewPkcs7Padding(8),
	}
}

func (p Pkcs5Padding) BlockSize() int {
	return p.pkcs7.BlockSize()
}

func (p Pkcs5Padding) Pad(buf []byte) []byte {
	return p.pkcs7.Pad(buf)
}

func (p Pkcs5Padding) Unpad(buf []byte) ([]byte, error) {
	return p.pkcs7.Unpad(buf)
}


aes.go
package crypt

import (
	"crypto/aes"
	"crypto/cipher"
	"errors"
)

const (
	Aes128KeySize = 128 / 8       // Size is 16 bytes.
	Aes192KeySize = 192 / 8       // Size is 24 bytes.
	Aes256KeySize = 256 / 8       // Size is 32 bytes.
	AesIvSize     = aes.BlockSize // Size is 16 bytes.
	AesBlockSize  = aes.BlockSize // Size is 16 bytes.
)

var (
	errAesPaddingBlockSizeMustBeAesBlockSize = errors.New("padding block size must be aes block size")
	errAesDataSizeMustBeMultipleOfBlockSize  = errors.New("data size must be multiple of block size")
	errAesIvLenMustBeBlockSize               = errors.New("iv length must equal to block size")
)

// The key must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256.
// The iv must be 16 bytes.
// The block size of padding must be 16 bytes.
//
// Can call HasError to see if it has an error.
func NewAesCbcEncrypter(key, iv []byte, padding Padding) AesBlockModeEncrypter {
	block, err := checkKeyIvPadding(key, iv, padding)
	if err != nil {
		return newAesBlockModeEncrypter(nil, nil, err)
	}
	return newAesBlockModeEncrypter(cipher.NewCBCEncrypter(block, iv), padding, nil)
}

// The key must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256.
// The iv must be 16 bytes.
// The block size of padding must be 16 bytes.
//
// Can call HasError to see if it has an error.
func NewAesCbcDecrypter(key, iv []byte, padding Padding) AesBlockModeDecrypter {
	block, err := checkKeyIvPadding(key, iv, padding)
	if err != nil {
		return newAesBlockModeDecrypter(nil, nil, err)
	}
	return newAesBlockModeDecrypter(cipher.NewCBCDecrypter(block, iv), padding, nil)
}

// The key must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256.
// The iv must be 16 bytes.
//
// Can call HasError to see if it has an error.
func NewAesCfbEncrypter(key, iv []byte) AesStream {
	block, err := checkKeyIv(key, iv)
	if err != nil {
		return newAesStream(nil, err)
	}
	return newAesStream(cipher.NewCFBEncrypter(block, iv), nil)
}

// The key must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256.
// The iv must be 16 bytes.
//
// Can call HasError to see if it has an error.
func NewAesCfbDecrypter(key, iv []byte) AesStream {
	block, err := checkKeyIv(key, iv)
	if err != nil {
		return newAesStream(nil, err)
	}
	return newAesStream(cipher.NewCFBDecrypter(block, iv), nil)
}

// The key must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256.
// The iv must be 16 bytes.
//
// Can call HasError to see if it has an error.
func NewAesOfb(key, iv []byte) AesStream {
	block, err := checkKeyIv(key, iv)
	if err != nil {
		return newAesStream(nil, err)
	}
	return newAesStream(cipher.NewOFB(block, iv), nil)
}

// The key must be either 16, 24, or 32 bytes to select AES-128, AES-192, or AES-256.
// The iv must be 16 bytes.
//
// Can call HasError to see if it has an error.
func NewAesCtr(key, iv []byte) AesStream {
	block, err := checkKeyIv(key, iv)
	if err != nil {
		return newAesStream(nil, err)
	}
	return newAesStream(cipher.NewCTR(block, iv), nil)
}

func checkKeyIv(key, iv []byte) (cipher.Block, error) {
	if len(iv) != aes.BlockSize {
		return nil, errAesIvLenMustBeBlockSize
	}
	return aes.NewCipher(key)
}

func checkKeyIvPadding(key, iv []byte, padding Padding) (cipher.Block, error) {
	if padding.BlockSize() != AesBlockSize {
		return nil, errAesPaddingBlockSizeMustBeAesBlockSize
	}
	return checkKeyIv(key, iv)
}

// It may has an error, call HasError to see it.
type AesBlockModeEncrypter struct {
	blockMode cipher.BlockMode
	padding   Padding
	err       error
}

func newAesBlockModeEncrypter(blockMode cipher.BlockMode, padding Padding, err error) AesBlockModeEncrypter {
	return AesBlockModeEncrypter{
		blockMode: blockMode,
		padding:   padding,
		err:       err,
	}
}

func (e AesBlockModeEncrypter) HasError() (error, bool) {
	return e.err, e.err != nil
}

// The result will not share the array of src.
func (e AesBlockModeEncrypter) Encrypt(src []byte) ([]byte, error) {
	if e.err != nil {
		return nil, e.err
	}
	buf := e.padding.Pad(src)
	e.blockMode.CryptBlocks(buf, buf)
	return buf, nil
}

// It may has an error, call HasError to see it.
type AesBlockModeDecrypter struct {
	blockMode cipher.BlockMode
	padding   Padding
	err       error
}

func newAesBlockModeDecrypter(blockMode cipher.BlockMode, padding Padding, err error) AesBlockModeDecrypter {
	return AesBlockModeDecrypter{
		blockMode: blockMode,
		padding:   padding,
		err:       err,
	}
}

func (d AesBlockModeDecrypter) HasError() (error, bool) {
	return d.err, d.err != nil
}

// The result will not share the array of src.
func (d AesBlockModeDecrypter) Decrypt(src []byte) ([]byte, error) {
	if d.err != nil {
		return nil, d.err
	}
	if len(src)%d.blockMode.BlockSize() != 0 {
		return nil, errAesDataSizeMustBeMultipleOfBlockSize
	}
	dst := make([]byte, len(src))
	d.blockMode.CryptBlocks(dst, src)
	return d.padding.Unpad(dst)
}

// It may has an error, call HasError to see it.
type AesStream struct {
	stream cipher.Stream
	err    error
}

func newAesStream(stream cipher.Stream, err error) AesStream {
	return AesStream{
		stream: stream,
		err:    err,
	}
}

func (s AesStream) HasError() (error, bool) {
	return s.err, s.err != nil
}

// The result will not share the array of src.
func (s AesStream) Crypt(src []byte) ([]byte, error) {
	if s.err != nil {
		return nil, s.err
	}
	dst := make([]byte, len(src))
	s.stream.XORKeyStream(dst, src)
	return dst, nil
}
